src/Storages/S3Queue/StorageS3Queue.cpp

```diff
@@ -174,83 +194,69 @@ Pipe StorageS3Queue::read(
+                        "To enable use setting `stream_like_engine_allow_direct_select`");
+    }
 
     if (mv_attached)
-        throw Exception(ErrorCodes::QUERY_NOT_ALLOWED, "Cannot read from StorageS3Queue with attached materialized views");
-
-    auto query_configuration = updateConfigurationAndGetCopy(local_context);
-
-    std::shared_ptr<StorageS3Source::IIterator> iterator_wrapper = createFileIterator(local_context, query_info.query);
-
-    auto read_from_format_info = prepareReadingFromFormat(column_names, storage_snapshot, supportsSubsetOfColumns(local_context), getVirtuals());
-
-    const size_t max_download_threads = local_context->getSettingsRef().max_download_threads;
-
-    return Pipe(std::make_shared<StorageS3QueueSource>(
-        read_from_format_info,
-        configuration.format,
-        getName(),
-        local_context,
-        format_settings,
-        max_block_size,
-        query_configuration.request_settings,
-        configuration.compression_method,
-        query_configuration.client,
-        query_configuration.url.bucket,
-        query_configuration.url.version_id,
-        query_configuration.url.uri.getHost() + std::to_string(query_configuration.url.uri.getPort()),
-        iterator_wrapper,
-        files_metadata,
-        after_processing,
-        max_download_threads));
-}
-
-SinkToStoragePtr StorageS3Queue::write(const ASTPtr &, const StorageMetadataPtr &, ContextPtr, bool)
-{
-    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Write is not supported by storage {}", getName());
-}
-
-void StorageS3Queue::truncate(const ASTPtr &, const StorageMetadataPtr &, ContextPtr, TableExclusiveLockHolder &)
-{
-    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Truncate is not supported by storage {}", getName());
-}
-
-NamesAndTypesList StorageS3Queue::getVirtuals() const
-{
-    return virtual_columns;
-}
+    {
+        throw Exception(ErrorCodes::QUERY_NOT_ALLOWED,
+                        "Cannot read from {} with attached materialized views", getName());
+    }
 
-bool StorageS3Queue::supportsPartitionBy() const
-{
-    return true;
+    Pipes pipes;
+    const size_t adjusted_num_streams = std::min<size_t>(num_streams, s3queue_settings->s3queue_processing_threads_num);
+    for (size_t i = 0; i < adjusted_num_streams; ++i)
+        pipes.emplace_back(createSource(column_names, storage_snapshot, query_info.query, max_block_size, local_context));
+    return Pipe::unitePipes(std::move(pipes));
 }
 
-void StorageS3Queue::startup()
+std::shared_ptr<StorageS3QueueSource> StorageS3Queue::createSource(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    ASTPtr query,
+    size_t max_block_size,
+    ContextPtr local_context)
 {
-    if (task)
-        task->holder->activateAndSchedule();
-}
+    auto configuration_snapshot = updateConfigurationAndGetCopy(local_context);
+    auto file_iterator = createFileIterator(local_context, query);
```

Comment:
I see we create a separate file oterator for each source. So all sources will iterate through the same files and coordinate through ZK. I think it will be more optimal to share single file iterator across all sources, it will reduce the amount of requests to ZK for the same files